python/mcap/mcap/_message_queue.py

```diff
@@ -65,34 +67,67 @@ def position(self) -> Tuple[int, Optional[int]]:
         return (self.item[1], self.item[2])
 
 
-class MessageQueue:
-    """A queue of MCAP messages and chunk indices.
+def _make_orderable(item: QueueItem, reverse: bool) -> _Orderable:
+    if isinstance(item, ChunkIndex):
+        return _ChunkIndexWrapper(item, reverse)
+    return _MessageTupleWrapper(item, reverse)
+
+
+class _MessageQueue(ABC):
+    @abstractmethod
+    def push(self, item: QueueItem):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def pop(self) -> QueueItem:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def __len__(self) -> int:
+        raise NotImplementedError()
 
-    :param log_time_order: if True, this queue acts as a priority queue, ordered by log time.
-        if False, ``pop()`` returns elements in insert order.
-    :param reverse: if True, order elements in descending log time order rather than ascending.
-    """
 
-    def __init__(self, log_time_order: bool, reverse: bool = False):
+class LogTimeOrderQueue(_MessageQueue):
+    def __init__(self, reverse: bool = False):
         self._q: List[_Orderable] = []
-        self._log_time_order = log_time_order
         self._reverse = reverse
 
     def push(self, item: QueueItem):
-        if isinstance(item, ChunkIndex):
-            orderable = _ChunkIndexWrapper(item, self._reverse)
-        else:
-            orderable = _MessageTupleWrapper(item, self._reverse)
-        if self._log_time_order:
-            heapq.heappush(self._q, orderable)
-        else:
-            self._q.append(orderable)
+        orderable = _make_orderable(item, self._reverse)
+        heapq.heappush(self._q, orderable)
+
+    def pop(self) -> QueueItem:
+        return heapq.heappop(self._q).item
+
+    def __len__(self) -> int:
+        return len(self._q)
+
+
+class InsertOrderQueue(_MessageQueue):
+    def __init__(self, reverse: bool = False):
+        self._q: Deque[_Orderable] = deque()
+        self._reverse = reverse
+
+    def push(self, item: QueueItem):
+        orderable = _make_orderable(item, self._reverse)
```

Comment:
why do you care about making an orderable if the collection does not use the comparison operator?

Reply:
Good point! I guess `log_time_order=False` with `reverse=True` doesn't really make sense? Should we do a `ValueError` in the factory function then?